### PR TITLE
Add support for depth-guided training

### PR DESCRIPTION
**Pull request by filipstrand**
Have not tried this yet so this will have to be evaluated before merged as a proper feature, but especially for subject-related LoRAs it feels like a lot of training could be wasted on modelling unnecessary details often present in the background. With DepthPro now natively supported, I thought scaling the loss according to the depth-map could potentially lead to faster convergence.

### CODE REVIEW

File: src/mflux/dreambooth/_example/train.json [hunk 1]
```diff
@@ -44,23 +44,28 @@
       "images": [
         {
           "image": "01.jpeg",
-          "prompt": "photo of sks dog"
+          "prompt": "photo of sks dog",
+          "use_depth": true
```
Comment (filipstrand) [inline src/mflux/dreambooth/_example/train.json:48]: The idea is to specify `use_depth` on a per-example basis

File: src/mflux/dreambooth/dataset/dataset.py [hunk 2]
```diff
@@ -61,19 +73,42 @@ def _create_examples(
...
-            mx.eval(pooled_prompt_embeds)
+            mx.eval(encoded_image, prompt_embeds, pooled_prompt_embeds)
+            if depth_map is not None:
+                mx.eval(depth_map)
 
         return examples
 
     @staticmethod
-    def _encode_image(vae: nn.Module, image_path: Path, width: int, height: int) -> mx.array:
+    def _create_and_encode_depth_map(
```
Comment (filipstrand) [inline src/mflux/dreambooth/dataset/dataset.py:88]: Depth map is created when the examples are generated before training

File: src/mflux/dreambooth/optimization/dreambooth_loss.py [hunk 3]
```diff
@@ -62,4 +62,18 @@ def _single_example_loss(flux: Flux1, config: RuntimeConfig, example: Example, r
         )
 
         # Construct the loss (derivation in src/mflux/dreambooth/optimization/_loss_derivation)
-        return (clean_image + predicted_noise - pure_noise).square().mean()
+        pixel_losses = (clean_image + predicted_noise - pure_noise).square()
```
Comment (filipstrand) [inline src/mflux/dreambooth/optimization/dreambooth_loss.py:65]: This loss kind of assumes that it make sense to simply encode the depth image and that the spatial information is more or less maintained in the lower resolution latent. (i.e this would not work if the latents were more abstract and did not retain any spatial layout)

File: src/mflux/dreambooth/optimization/dreambooth_loss.py [hunk 4]
```diff
@@ -62,4 +62,18 @@ def _single_example_loss(flux: Flux1, config: RuntimeConfig, example: Example, r
...
+
+    @staticmethod
+    def scale_loss_with_depth_map(depth_map: mx.array, pixel_losses: mx.float16) -> mx.float16:
+        # 1. Normalize the depth map to [0, 1]
+        depth_map_normalized = (depth_map - mx.min(depth_map)) / (mx.max(depth_map) - mx.min(depth_map) + 1e-8)
+
+        # 2. Normalize weights to sum to 1
+        weights = depth_map_normalized / (mx.sum(depth_map_normalized) + 1e-8)
+        weighted_loss = mx.sum(pixel_losses * weights)
+        return weighted_loss
```
Comment (filipstrand) [inline src/mflux/dreambooth/optimization/dreambooth_loss.py:79]: This loss kind of assumes that it make sense to simply encode the depth image and that the spatial information is more or less maintained in the lower resolution latent. (i.e this would not work if the latents were more abstract and did not retain any spatial layout)

**Comment by anthonywu** [comment]
Looking forward to this feature, I definitely have observed unwanted artifacts that are unmistakably from the training set.

Related, the Mac Finder has the right-click remove background Service, which I think is exactly what you integrating as part of the workflow here. Really good addition.

**Comment by filipstrand** [comment]
> Looking forward to this feature, I definitely have observed unwanted artifacts that are unmistakably from the training set.
> 
> Related, the Mac Finder has the right-click remove background Service, which I think is exactly what you integrating as part of the workflow here. Really good addition.

Thanks! Yes, now that we have the model in place I kinda want to find more use-cases for it :)  The model itself seem to perform really well in my testing (and is also really fast), handling hair and edges really well! I tried the depthAnything model before committing to this one (via the standard Pytorch implementation), but typically got a softer depth map with less detail.

Just got around to doing a training run today and my initial results were not that different from training without the depth map... Will do some more testing of this feature over the coming week. Maybe it is worth having some kind of scale that says how much of the depth area to reward/penalise in the loss (e.g to express that the "near-white" is a lot more important than grey etc. in a non-linear way)

**Comment by filipstrand** [comment]
@claude Can you summarise what was done in d232f91b3612acc4df03358efa2abc7d79db54e9?
